src/php/parser/Parser.php

```diff
@@ -0,0 +1,545 @@
+            }
+        }
+
+        throw new RuntimeException("must not be here!");
+    }
+
+    private function moveEval(EvalAction $action): Action
+    {
+        $expression = $action->getExpression();
+        $position = $action->getPosition();
+        $asts = $action->getAsts();
+        $error = $action->getError();
+        $continuations = $action->getContinuations();
+        $eof = $position >= $this->inputLength;
+
+        switch ($expression->getType()) {
+            case ExpressionType::ANY_CHAR:
+            {
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new WildcardError()));
+                } else {
+                    $char = $this->codePointAt($this->input, $position, $this->codePointMaxLength);
+                    $len = strlen($char);
+
+                    $matchResult = $this->accept($position + $len, ASTs::asts(new Char($char, $position), $asts), $error);
+                }
+
+                return $this->applyNext($continuations, $matchResult);
+            }
+            case ExpressionType::ALT:
+            {
+                $expressions = Expression::asAltExpression($expression)->getExpressions();
+
+                if ($expressions->isEmpty()) {
+                    return $this->applyNext($continuations, $this->reject($error));
+                } else {
+                    $cons = Continuations::cons(new AltContinuation($expressions->tail(), $position, $asts), $continuations);
+                    return $this->evalNext($expressions->head(), $position, $asts, $error, $cons);
+                }
+            }
+            case ExpressionType:: AND:
+            {
+                $cons = Continuations::cons(new AndContinuation($position, $asts, $error), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::NOT:
+            {
+                $cons = Continuations::cons(new NotContinuation($position, $asts, $error), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::VOID:
+            {
+                $cons = Continuations::cons(new VoidContinuation($asts), $continuations);
+                return $this->evalNext($expression->getExpression(), $position, new ASTs(), $error, $cons);
+            }
+            case ExpressionType::CHAR:
+            {
+                $expression = CharExpression::asCharExpression($expression);
+
+                if ($eof) {
+                    $matchResult = $this->reject($this->updateError($error, $position, new CharacterError($expression)));
+                } else {
+                    $expected = Expression::asCharExpression($expression)->getChar();
+                    $actual = $this->codePointAt($this->input, $position, $this->codePointMaxLength);
```

Comment:
There's no need to decode the code point here. You can simply do:

```php
$matches = mb_strcut($this->input, $position, strlen($expected)) == $expected;
```

Reply:
resolved in dd422d6a91326df014491d028a9265a75cbc002f